src/imap/idle.rs

```diff
@@ -91,17 +94,25 @@ impl Imap {
                                 }
                             }
                         }
-                        match handle.done().await {
-                            Ok(session) => {
+                        // if we can't properly terminate the idle
+                        // protocol let's break the connection.
+                        let res =
+                            async_std::future::timeout(Duration::from_secs(15), { handle.done() });
+                        match res.await {
+                            Ok(Ok(session)) => {
                                 *self.session.lock().await = Some(Session::Secure(session));
                             }
-                            Err(err) => {
+                            Ok(Err(err)) => {
                                 // if we cannot terminate IDLE it probably
                                 // means that we waited long (with idle_wait)
                                 // but the network went away/changed
                                 self.trigger_reconnect();
                                 return Err(Error::IdleProtocolFailed(err));
                             }
+                            Err(err) => {
```

Comment:
Maybe introduce a temporary variable and check for timeout error first, so you don't have to match on `Ok(Ok` which is hard to read.